File: core-platform/services/core-service/src/main/java/com/dev/core/service/leave/LeaveBalanceService.java
package com.dev.core.service.leave;


import java.util.List;

import com.dev.core.model.leave.LeaveBalanceDTO;
import com.dev.core.model.leave.MinimalLeaveBalanceDTO;

public interface LeaveBalanceService {

    LeaveBalanceDTO getBalance(Long employeeId, Long leaveTypeId, Integer year);

    List<LeaveBalanceDTO> getAllBalances(Long employeeId, Integer year);

    List<MinimalLeaveBalanceDTO> getMinimalBalances(Long employeeId, Integer year);

    void initializeYearlyBalance(Long employeeId, Integer year);

    void accrueMonthlyEarnedLeaves(Integer year, Integer month);

    void carryForwardBalances(Integer year);
}

File: core-platform/apps/core-webapp/src/services/leaveBalance.service.ts
import axiosInstance from "../axiosInstance";
import type {
  LeaveBalanceDTO,
  MinimalLeaveBalanceDTO,
} from "../types/leave.types";

const LEAVE_BALANCES_API_BASE = "/leave-balances";

export const leaveBalanceService = {
  // Get balance for a specific employee, leave type, and year
  getBalance: async (
    employeeId: number,
    leaveTypeId: number,
    year: number,
  ): Promise<LeaveBalanceDTO> => {
    const response = await axiosInstance.get<LeaveBalanceDTO>(
      `${LEAVE_BALANCES_API_BASE}/${employeeId}/${leaveTypeId}/${year}`,
    );
    return response.data;
  },

  // Get all balances for an employee in a specific year
  getAllBalances: async (
    employeeId: number,
    year: number,
  ): Promise<LeaveBalanceDTO[]> => {
    const response = await axiosInstance.get<LeaveBalanceDTO[]>(
      `${LEAVE_BALANCES_API_BASE}/${employeeId}/year/${year}`,
    );
    return response.data;
  },

  // Get minimal balances for an employee in a specific year
  getMinimalBalances: async (
    employeeId: number,
    year: number,
  ): Promise<MinimalLeaveBalanceDTO[]> => {
    const response = await axiosInstance.get<MinimalLeaveBalanceDTO[]>(
      `${LEAVE_BALANCES_API_BASE}/minimal/${employeeId}/year/${year}`,
    );
    return response.data;
  },

  // Initialize yearly balance for an employee
  initializeYearlyBalance: async (
    employeeId: number,
    year: number,
  ): Promise<void> => {
    await axiosInstance.post(
      `${LEAVE_BALANCES_API_BASE}/initialize/${employeeId}/${year}`,
    );
  },
};

export default leaveBalanceService;
